### PR TITLE
#1955 C# Web API Controller : Generator does not handle "required" fi…

### PR DESCRIPTION
**Pull request by nikobiker12**
Can you review the PR please ?
I added two parameters in the settings to sets whether attributes must be generated or not and which type of attribute must be generated.
I refactored the unit tests by using a swagger document "manually" generated instead of using an generated one

### CODE REVIEW

File: src/NSwag.CodeGeneration.CSharp/SwaggerToCSharpControllerGeneratorSettings.cs [hunk 1]
```diff
@@ -24,6 +24,8 @@ public SwaggerToCSharpControllerGeneratorSettings()
             UseCancellationToken = false;
             AspNetNamespace = "System.Web.Http";
             RouteNamingStrategy = CSharpControllerRouteNamingStrategy.None;
+            UseModelValidationAttributes = false;
+            RequiredAttributeType = "System.ComponentModel.DataAnnotations.Required";
```
Comment (RSuter) [inline src/NSwag.CodeGeneration.CSharp/SwaggerToCSharpControllerGeneratorSettings.cs]: ASP.NET Core needs the BindRequiredAttribute to make parameters required not the RequiredAttribute

**Comment by nikobiker12** [comment]
I modified the value of the RequiredAttributeType to BindRequired. Is it Ok ?

**Comment by RSuter** [comment]
I’m wondering if we should just drop & remove web api support and concentrate completely on aspnetcore? I’d say yes...

/cc @rogersillito

**Comment by optuuttila** [comment]
In my view all focus should be put to aspnetcore. 

That's definitely the direction where things are going and allthough there will still be a lot of older webapi users for the time being I think this would be yet another reason for them to just(*) migrate to adpnetcore.

* Been there so I know it's not that straightforward, but hey, that's why we have strangler patterns etc. that come to rescue.

**Comment by rogersillito** [comment]
We need to supply a controller interface for a client that will likely still need Web API support, so I would vote to retain support for that right now.   Is there something specific about this PR that's problematic for supporting both?

File: src/NSwag.CodeGeneration.CSharp/SwaggerToCSharpControllerGeneratorSettings.cs (hunk 1 above)
Comment (rogersillito) [reply src/NSwag.CodeGeneration.CSharp/SwaggerToCSharpControllerGeneratorSettings.cs]: I just read a post which may be useful here: [have BindAttribute behave as BindRequiredAttribute in ASP.NET Core MVC](https://www.strathweb.com/2017/12/required-and-bindrequired-in-asp-net-core-mvc) - see _"But I don’t want to mix BindRequiredAttribute and RequiredAttribute"_

**Comment by RicoSuter** [comment]
Ah man, it seems that nswag will support “legacy” framework much longer than the asp team does 🙂

Ok, ill add a new IsAspNetCore setting, remove the namespace setting (as it can be inferred) and then all controller gen PRs have to align with that and check the new flag - they must support at least aspnetcore but must not break web api

**Comment by RicoSuter** [comment]
Added PR with new ControllerTarget setting: https://github.com/RicoSuter/NSwag/pull/2064

File: src/NSwag.CodeGeneration.CSharp/Models/CSharpControllerTemplateModel.cs [hunk 2]
```diff
@@ -77,6 +77,12 @@ public class CSharpControllerTemplateModel : CSharpTemplateModelBase
         /// <summary>Gets a value indicating whether to allow adding cancellation token.</summary>
         public bool UseCancellationToken => _settings.UseCancellationToken;
 
+        /// <summary>Gets a value indicating whether to allow adding model validation attributes</summary>
+        public bool UseModelValidationAttributes => _settings.UseModelValidationAttributes;
+
+        /// <summary>Gets the type of the attribte used to specify a parameter as required.</summary>
+        public string RequiredAttributeType => _settings.RequiredAttributeType;
```
Comment (RicoSuter) [inline src/NSwag.CodeGeneration.CSharp/Models/CSharpControllerTemplateModel.cs]: Why do we need a setting for that? Can't we just check the ControllerTarget (AspNet/AspNetCore) and "hardcode" the full attribute name? Don't see a scenario where one would choose another type...

File: src/NSwag.CodeGeneration.CSharp/SwaggerToCSharpControllerGeneratorSettings.cs [hunk 3]
```diff
@@ -53,5 +55,11 @@ public string GetRouteName(SwaggerOperation operation)
 
         /// <summary>Gets or sets the strategy for naming routes (default: CSharpRouteNamingStrategy.None).</summary>
         public CSharpControllerRouteNamingStrategy RouteNamingStrategy { get; set; }
+
+        /// <summary>Gets or sets a value indicating whether to allow adding model validation attributes</summary>
+        public bool UseModelValidationAttributes { get; set; }
```
Comment (RicoSuter) [inline src/NSwag.CodeGeneration.CSharp/SwaggerToCSharpControllerGeneratorSettings.cs]: Maybe rename to GenerateModelValidationAttributes or GenerateValidationAttributes ?

**Comment by RicoSuter** [comment]
Please update your branch with master branch (fix conflicts)

File: src/NSwag.CodeGeneration.CSharp/Models/CSharpControllerTemplateModel.cs (hunk 2 above)
Comment (RicoSuter) [reply src/NSwag.CodeGeneration.CSharp/Models/CSharpControllerTemplateModel.cs]: Implement it the same way as latest version of the AspNetNamespace property?

File: src/NSwag.CodeGeneration.CSharp/SwaggerToCSharpControllerGeneratorSettings.cs (hunk 3 above)
Comment (nikobiker12) [reply src/NSwag.CodeGeneration.CSharp/SwaggerToCSharpControllerGeneratorSettings.cs]: I ve renamed to GenerateModelValidationAttributes

**Comment by nikobiker12** [comment]
@RicoSuter 
I ve juste took into account your remarks and merge the master branch
Sorry for the delay

File: src/NSwag.CodeGeneration.CSharp/Models/CSharpControllerTemplateModel.cs (hunk 2 above)
Comment (nikobiker12) [reply src/NSwag.CodeGeneration.CSharp/Models/CSharpControllerTemplateModel.cs]: done

**Comment by RicoSuter** [comment]
v12.2.2

**Pull request merged**